### PR TITLE
Readthedocs Implementation

### PR DESCRIPTION
**Pull request by tomschewski**
PR includes:

- Updated README
- Docs folder containing the files necessary for readthedocs functionality

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,56 +1,65 @@
...
+Experanto is a Python package designed for interpolating recordings and stimuli in neuroscience experiments. It enables users to load single or multiple experiments and create efficient dataloaders for machine learning applications.
 
-- Single frame or sequence access
-```python
-item = dat[10]
-sequence = dat[10:100]
+## Features
+- **Flexible Data Loading**: Load and process multiple experiments with ease.
+- **Neuroscience-Focused**: Designed for handling recordings and stimuli in neuroscience research.
+- **Seamless Machine Learning Integration**: Create dataloaders for ML models effortlessly.
```
Comment (fabiansinz) [inline README.md:8]: Can you please tone down the AI-speak here? Feels like the overboarding ChatGPT language.

**Review by fabiansinz — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (tomschewski) [reply README.md:8]: I just removed this passage, I believe the initial Introduction is sufficient.

File: docs/source/concepts/demo_dataset.rst [hunk 2]
```diff
@@ -0,0 +1,119 @@
+
+.. _loading_dataset:
+
+Loading a Dataset Object
+========================
+
+Dataset objects are the core of the Experanto library. They extend single experiments by adding functionality that allows them to be used directly as **dataloaders** for machine learning tasks.
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:7]: Are they the core? :D I am not sure whether I would name interpolators or datasets the core of experanto ... I am also not feeling so comfortable with "extend single experiments" and "used as dataloaders". My suggestion might also not perfect and can be discussed!
```suggestion
Dataset objects organize experimental data (from `Experiment` class) for machine learning tasks, offering project-specific and configurable access for training and evaluation. They often serve as a source for creating dataloaders.
```

File: docs/source/concepts/demo_dataset.rst [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+-------------------------------
+Dataset objects provide several essential features:
+
+- **Sampling Rate**: Defines the frequency of equally spaced interpolation times across the entire experiment. This ensures consistency in temporal data alignment.
+- **Chunk Size**: Determines the number of values returned when calling the ``__getitem__`` method. This is crucial for deep learning models utilizing **3D convolutions over time**, as single elements or small chunk sizes would be insufficient for meaningful temporal patterns.
+- **Modality Configuration**: Specifies the details of the interpolation, including:
+
+  - The **interpolation method** used.
+  - **Conditions** that the data must fulfill.
+  - **Transforms** applied to the data (e.g., normalization, resizing, cropping, greyscale conversion).
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:19]:
```suggestion
  - **Transformations** applied to the data (e.g., normalization, resizing, cropping, greyscale conversion).
```

File: docs/source/concepts/demo_dataset.rst [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    import sys
+    from experanto.datasets import ChunkDataset
+    from torch.utils.data import DataLoader
+    import numpy as np
+    from tqdm import tqdm
+    import matplotlib.pyplot as plt
+    from collections import OrderedDict
+
+    from omegaconf import OmegaConf, open_dict
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:35]: sys, np, tqdm, plt, OrderedDict, open_dict are not used in this code block

File: docs/source/concepts/demo_dataset.rst [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    train_dataset = ChunkDataset(root_folder=f'{root_folder}/experiment_951980471_train', global_sampling_rate=sampling_rate,
+            global_chunk_size=chunk_size, modality_config = selected_modalities)
+
+This configuration ensures that:
+
+- **Screen data** is preprocessed with normalization, resizing, cropping, and greyscale conversion.
+- **Response data** undergoes standardization and nearest-neighbor interpolation.
+
+Other modalities can be defined in the same manner as **Responses**.
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:61]:
```suggestion
Other modalities can be defined in the same manner.
```

File: docs/source/concepts/demo_dataset.rst [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    sampling_rate = 60
+    chunk_size = 60 # since we also use video data we always use chunks of images to also consider temporal developements
+
+    train_dataset = ChunkDataset(root_folder=f'{root_folder}/experiment_951980471_train', global_sampling_rate=sampling_rate,
+            global_chunk_size=chunk_size, modality_config = selected_modalities)
+
+This configuration ensures that:
+
+- **Screen data** is preprocessed with normalization, resizing, cropping, and greyscale conversion.
+- **Response data** undergoes standardization and nearest-neighbor interpolation.
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:59]: Because both are per default enabled, right? I would mention that then, otherwise it might confuse.
```suggestion
- **Response data** undergoes standardization and nearest-neighbor interpolation (per default).
```

File: docs/source/concepts/demo_dataset.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    root_folder = '../data/allen_data'
+    sampling_rate = 60
+    chunk_size = 60 # since we also use video data we always use chunks of images to also consider temporal developements
+
+    train_dataset = ChunkDataset(root_folder=f'{root_folder}/experiment_951980471_train', global_sampling_rate=sampling_rate,
+            global_chunk_size=chunk_size, modality_config = selected_modalities)
+
+This configuration ensures that:
+
+- **Screen data** is preprocessed with normalization, resizing, cropping, and greyscale conversion.
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:58]: Normalization is applied per default, right?
```suggestion
- **Screen data** is preprocessed with normalization (per default), resizing, cropping, and greyscale conversion.
```

Just a question for myself: interpolation.rescale_size adds cropping or how is cropping achieved in this code block?

File: docs/source/concepts/demo_dataset.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    from torch.utils.data import DataLoader
+    import numpy as np
+    from tqdm import tqdm
+    import matplotlib.pyplot as plt
+    from collections import OrderedDict
+
+    from omegaconf import OmegaConf, open_dict
+    from experanto.configs import DEFAULT_CONFIG as cfg
+
+    cfg.dataset.modality_config.screen.transforms.Resize.size = [144,144]
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:38]:
```suggestion
    cfg.dataset.modality_config.screen.transforms.Resize.size = [144, 144] 
```

File: docs/source/concepts/demo_dataset.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    for key in sample.keys():
+        print(f'This is shape {sample[key].shape} for modality {key}')
+
+This will output something like:
+
+.. code-block:: text
+
+    dict_keys(['screen', 'responses'])
+    This is shape torch.Size([1, 32, 144, 144]) for modality screen
+    This is shape torch.Size([32, 12]) for modality responses
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:84]: This should match the config above, no? So 60 instead of 32?

File: docs/source/concepts/demo_dataset.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+Defining DataLoaders
+---------------------
+Once the dataset is verified, we can define **DataLoader** objects for training or other purposes. This allows easy batch processing during training:
+
+.. code-block:: python
+
+    # Define a DataLoader for the training set
+    data_loader['train'] = DataLoader(train_dataset, batch_size=32, shuffle=True)
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:93]: This does not work without specifying data_loader before. So I would just use `train_data_loader` or so to not confuse.

File: docs/source/concepts/demo_dataset.rst [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+----------------------------------
+To confirm that the **DataLoader** works as expected, we can iterate over it and inspect the batch data. For example, to check the shapes of the data in each batch:
+
+.. code-block:: python
+
+    # Interpolation showcase using the data_loaders
+    for batch_idx, batch_data in enumerate(data_loaders['train']):
+        # batch_data is a dictionary with keys 'screen', 'responses'
+        screen_data = batch_data['screen']
+        responses = batch_data['responses']
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:105]: I feel either `screen` or `responses_data` would be more consistent.

File: docs/source/concepts/demo_dataset.rst [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        
+        # Print or inspect the batch
+        print(f"Batch {batch_idx}:")
+        print("Screen Data:", screen_data.shape)
+        print("Responses:", responses.shape)
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:110]: Same as above

File: docs/source/concepts/demo_dataset.rst [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        break
+
+This will output something like:
+
+.. code-block:: text
+
+    Batch 0:
+    Screen Data: torch.Size([15, 1, 32, 144, 144])
+    Responses: torch.Size([15, 32, 12])
```
Comment (reneburghardt) [inline docs/source/concepts/demo_dataset.rst:119]: Either I am absolutely confused or the shape should be [32, 60, 12] according to the earlier configurations, no?

File: docs/source/concepts/installation.rst [hunk 14]
```diff
@@ -0,0 +1,15 @@
+Installation
+============
+
+**experanto** is available for Python 3.9
+
+The package works ontop of jupyter/datascience-notebooks but minimum requirements are:
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/installation.rst:6]: This should be correct :: indicates a code block in the next lines

File: docs/source/concepts/installation.rst [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+- ``numpy``
+- ``pandas``
+- ``torch``
+
+to install the package, clone it into a local repository and run::
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/installation.rst:12]: This should be correct, :: comes before a code block in rst syntax.

File: docs/source/concepts/installation.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/installation.rst:6]: Corrected this.

File: docs/source/concepts/getting_started.rst [hunk 16]
```diff
@@ -0,0 +1,10 @@
+Use cases for experanto
+=======================
+
+**experanto** help aligning temporal data from neuralscience experiments. It can handle visual stimuli, behaviours and reponses from neurons.
+
+It fixes usual issues such as:
+
+- Different sampling rates for different measurements
+- Differing formats for datatypes such as videos and images in the same measurements
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/getting_started.rst:9]: Well I implemented it but it is not pushed to main yet. (Img's as .npy and videos as .npy or .mp4)
Corrected poor phrasing.

File: docs/source/concepts/getting_started.rst [hunk 17]
```diff
@@ -0,0 +1,10 @@
+Use cases for experanto
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/getting_started.rst:1]: Don't know why Cases should be uppercase but fixed the E.

File: README.md [hunk 18]
```diff
@@ -1,56 +1,58 @@
...
 ```
+(Repository: [sensorium_2023](https://github.com/ecker-lab/sensorium_2023))
 
-## Example for meta.yaml
+Ensure you replace `/path_to/` with the actual path to the cloned repositories.
 
-```
-modality: images
-```
+## Project structure
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply README.md:31]: No, this is an artifact from a previous merge with Konsti's PR.

File: docs/source/index.rst [hunk 19]
```diff
@@ -0,0 +1,26 @@
+**experanto**
+======================================
+
+**experanto** is a Python package to interpolate recordings and stimuli of neuroscience experiments.
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/index.rst:4]: Good Idea

File: README.md [hunk 20]
```diff
@@ -1,56 +1,58 @@
...
-  running_wheel/
-    meta.yaml
-    timestamps.npz
-  multiunit/
-    meta.yaml
-    timestamps.npz
-  poses/
-    meta.yaml
-    timestamps.npz
+Additionally, clone the `sensorium_2023` repository and add the directory to the Python path using:
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply README.md:22]: Absolutely agree, I use it myself that way. Will open PR to sensorium2023 that includes setup.py and necessary tools for pip -e install. And will update this once it is accepted or discussed differently.

File: docs/source/concepts/demo_configs.rst [hunk 21]
```diff
@@ -0,0 +1,111 @@
+Dataset and Dataloader Configuration
+====================================
+
+This section describes the dataset configuration used to load experiments into the Experanto dataloaders. It includes global settings, modality-specific configurations, and dataloader parameters.
+
+Default YAML Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/demo_configs.rst:9]: I agree that the codeblock should be secondary but personally I didn't use the omegaconf myself until recently due to a lack of documentation on it inside of experanto. Hence I think it should be keept just to make users familiar with how we handle our cfg.

File: docs/source/concepts/demo_data.rst [hunk 22]
```diff
@@ -0,0 +1,49 @@
+.. _generating_sample_data:
+
+Generating Sample Data
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/demo_data.rst:3]: Agree, I changed it to "Experanto data format".

File: docs/source/concepts/demo_data.rst [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+--------------
+A typical experiment follows this structured hierarchy:
+
+.. code-block:: text
+
+    experiment/
+      ├── screen/
+      │   ├── data/
+      │   │   ├── 00000.npy
+      │   │   ├── 00001.mp4
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/demo_data.rst:25]: Good suggestion.

File: docs/source/concepts/demo_experiment.rst [hunk 24]
```diff
@@ -0,0 +1,72 @@
+.. _loading_single_experiment:
+
+Loading a Single Experiment
+===========================
+
+To load an experiment, we use the **Experiment** class. This is particularly useful for testing whether the formatting and interpolation behave as expected before loading multiple experiments into dataset objects.
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/demo_experiment.rst:6]: I agree that the dataloaders and datasets are the main point of experanto for users. However I personally really liked progressing from experiments to dataloaders as I was confident that the interpolation worked as intended.
I personally think it's fine to recommend this to new users as a way to get familiar and test the library.

File: docs/source/concepts/demo_dataset.rst [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+========================
+
+Dataset objects are the core of the Experanto library. They extend single experiments by adding functionality that allows them to be used directly as **dataloaders** for machine learning tasks.
+
+Key Features of Dataset Objects
+-------------------------------
+Dataset objects provide several essential features:
+
+- **Sampling Rate**: Defines the frequency of equally spaced interpolation times across the entire experiment. This ensures consistency in temporal data alignment.
+- **Chunk Size**: Determines the number of values returned when calling the ``__getitem__`` method. This is crucial for deep learning models utilizing **3D convolutions over time**, as single elements or small chunk sizes would be insufficient for meaningful temporal patterns.
```
[the remark this answers is not in the document]
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:14]: Better phrasing.

File: docs/source/concepts/demo_dataset.rst (hunk 2 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:7]: I don't know myself about the core part but I agree that it I phrased it rather extreme. I don't have complains about your version.

File: docs/source/concepts/demo_dataset.rst (hunk 5 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:61]: Why remove this? In case of multiple screens? Because behaviors have the same as Responses so it it not more precise?

File: docs/source/concepts/demo_dataset.rst (hunk 6 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:59]: Yeah especially with more interpolators being added in the future.

File: docs/source/concepts/demo_dataset.rst (hunk 7 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:58]: I think in this example I don't apply cropping.  I believe I used a different config before which included center cropping.

File: docs/source/concepts/demo_dataset.rst (hunk 9 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:84]: Yes

File: docs/source/concepts/demo_dataset.rst (hunk 10 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:93]: You're right, I copied this without including the definition of data_loader.

File: docs/source/concepts/demo_dataset.rst (hunk 11 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:105]: changed both to _data

File: docs/source/concepts/demo_dataset.rst (hunk 13 above)
Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:119]: Yes you're right.

File: docs/source/concepts/demo_dataset.rst (hunk 5 above)
Comment (reneburghardt) [reply docs/source/concepts/demo_dataset.rst:61]: I felt its a bit misleading, because in general its possible to introduce modalities that are close to screen or close to responses or even not close to both (probably with additional tweaking). It just happens to be the case for us that behavior is more similar to responses

Comment (tomschewski) [reply docs/source/concepts/demo_dataset.rst:61]: Alright I will change it so that it states that modalities with the same format can be simply added and other modalities which differ have to be implemented manually.
